README.md
Čtečka SPZ <br>
Zoneminder <br>
Droidcam/ipcam (otestovat, která bude v praxi lepší) <br>
dostat se do bodu, kdy camera přečte spz a zapíše jí někam (bod odkud se program dá připojit na jiné databáze a různě využít.)



#potřeba přepsat
<h2> Linux <h2>
pip install python3.6 <br>
//needed python is usualy installed by default <br>
pycharm from app store
<h3> OpenCV <h3>
sudo apt update <br>
sudo apt install python3-opencv
<h3> Imultis <h3>
sudo pip install imutils
<h3> EasyOCR <h3>
pip install easyocr <br>
pip install git+git://github.com/jaidedai/easyocr.git

  
všechno integrovat pycharm


changes.md
from datetime import datetime
import cv2
from matplotlib import pyplot as plt
import numpy as np
import imutils
import easyocr

#define video
video = cv2.VideoCapture('/home/student/PycharmProjects/pythonProject/video1.mp4')
#frames = 0
#while(video.isOpened()):
#read frame by frame
#define variables
count = 0
frames = 1
skip_frames = 30
#loop function for using only frames % 31
while video.isOpened():
    ret, frame = video.read()
    if ret and frames % skip_frames == 0:
        # save frame as JPEG file
        cv2.imwrite("frame%d.jpg" % count, frame)
        #my trie to save image as variable not a .jpg
        #img = plt.Image.open(image)
        #add plus one to variables to use another frame and save next image with different name
        count += 1
        frames += 1

    else:
        frames += 1
video.release()
cv2.destroyAllWindows()
#save image to variable
#image = pil.Image.open('image')
#loop whole code to read every saved frame
number = 0
for number in range(count):
    #define variable img for used image
    img = cv2.imread('frame%d.jpg' %number)
    #img = cv2.imread('/home/jouadamis/PycharmProjects/pythonProject/image1.jpg')
    #converting picture to gray
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    #using bilateral Filter that reduces noise
    bfilter = cv2.bilateralFilter(gray, 11, 17, 17)
    #detecting edges
    edged = cv2.Canny(bfilter, 30, 200)
    #find edges
    keypoints = cv2.findContours(edged.copy(), cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
    #simplifying return of contours
    contours = imutils.grab_contours(keypoints)
    #return of top ten kontours
    contours = sorted(contours, key=cv2.contourArea, reverse=True)[:10]
    #filtring(looping) thrue contours to find number plate shape(rectangle)
    location = None
    for contour in contours:
        approx = cv2.approxPolyDP(contour, 10, True)
        if len(approx) == 4:
            location = approx
            break
    mask = np.zeros(gray.shape, np.uint8)
    #finding numberplate
    new_image = cv2.drawContours(mask, [location], 0,255, -1)
    #overlaying everything except numberplate
    new_image = cv2.bitwise_and(img, img, mask=mask)
    #storing coordinates in variables x,y
    (x,y) = np.where(mask==255)
    #storing top left point(min) and bottom right point(max)
    (x1, y1) = (np.min(x), np.min(y))
    (x2, y2) = (np.max(x), np.max(y))
    #getting variable croped_image
    cropped_image = gray[x1:x2+1, y1:y2+1]
    #define in what leanguage to read in
    reader = easyocr.Reader(['en'])
    #read text from variable cropped_image
    result = reader.readtext(cropped_image)
    f = open("file.txt", "a")
    f.write(str(result)  + " \n")
    f.close()
    #Render Result
    text1 = result[0][-2]
    text2 = result[1][-2]
    text = text1 + " " + text2
    #define font on picture
    font = cv2.FONT_HERSHEY_SIMPLEX
    #putting text in picture
    res = cv2.putText(img, text=text , org=(approx[0][0][0], approx[1][0][1]+60), fontFace=font, fontScale=1, color=(0,255,0), thickness=2, lineType=cv2.LINE_AA)
    #open txt file
    f = open("demofile.txt", "a")
    #define date + time to variable time
    time = datetime.now()
    #write in file SPZ and date+time
    f.write(text +"    " + str(time) + " \n")
    #close file
    f.close()
    #drawing rectangle in picture
    res = cv2.rectangle(img, tuple(approx[0][0]), tuple(approx[2][0]), (0,255,0),3)
    #save image
    plt.imsave('image_new%d.jpg' %number ,  cv2.cvtColor(res, cv2.COLOR_BGR2RGB))
    #plt.imsave('image_new1.jpg', cv2.cvtColor(res, cv2.COLOR_BGR2RGB))
    
    
    
    
    
    
    
    import cv2

# Opens the Video file
cap = cv2.VideoCapture('video1.mp4')
i = 0
while (cap.isOpened()):
    ret, frame = cap.read()
    if ret and frames % skip_frames == 0:
    if ret == False:
        break
    cv2.imwrite('kang' + str(i) + '.jpg', frame)
    i += 1

cap.release()
cv2.destroyAllWindows()
    


lpdb/alpr/rtsp.py
import cv2
from PIL import Image
from dotenv import load_dotenv
load_dotenv()
import os
import ultimateAlprSdk
import json
import requests
from datetime import datetime, timedelta
import base64
from io import BytesIO

# RTSP_PW = os.environ.get("RTSP_PW")
# RTSP_URL = "http://" + RTSP_PW + "@192.168.0.154:8080"

RTSP_URL = "http://192.168.0.154:8080"


JSON_CONFIG = {
    "debug_level": "info",
    "debug_write_input_image_enabled": False,
    "debug_internal_data_path": ".",

    "num_threads": -1,
    "gpgpu_enabled": True,
    "max_latency": -1,

    "klass_vcr_gamma": 1.5,

    "detect_roi": [0, 0, 0, 0],
    "detect_minscore": 0.1,

    "car_noplate_detect_min_score": 0.8,

    "pyramidal_search_enabled": True,
    "pyramidal_search_sensitivity": 0.28,
    "pyramidal_search_minscore": 0.3,
    "pyramidal_search_min_image_size_inpixels": 800,

    "recogn_minscore": 0.3,
    "recogn_score_type": "min",

    "assets_folder": "ultimateAlpr/assets",
    # Update JSON options using values from the command args
    "charset": "latin",
    "car_noplate_detect_enabled": False,
    "ienv_enabled": True,
    "openvino_enabled": True,
    "openvino_device": "CPU",
    "klass_lpci_enabled": False,
    "klass_vcr_enabled": False,
    "klass_vmmr_enabled": False,
    "klass_vbsr_enabled": False,
    "license_token_file": "",
    "license_token_data": ""
    }

ultimateAlprSdk.UltAlprSdkEngine_init(json.dumps(JSON_CONFIG))

vcap = cv2.VideoCapture(RTSP_URL)

recent_visits = {}
last_plate = ""
last_plate_count = 0

def add_plate(plate):
  print("Adding plate!", flush=True)
  t = datetime.today()

  # Resize image
  base_width = 1000
  wpercent = (base_width/float(image.size[0]))
  hsize = int((float(image.size[1])*float(wpercent)))
  scaled_image = image.resize((base_width, hsize), Image.ANTIALIAS)

  # Encode image to b64
  buffered = BytesIO()
  scaled_image.save(buffered, format="PNG")
  b64_image = base64.b64encode(buffered.getvalue())

  requests.post("http://backend:9000/api/visits", data = {'plate': plate, 'timestamp': t, 'image': b64_image})
  recent_visits[plate] = t

while(1):
  ret, frame = vcap.read()
  #frame = cv2.Canny(frame,100,200)
  # Convert from opencv image format to PIL image format
  

  image = Image.fromarray(cv2.cvtColor(frame, cv2.COLOR_BGR2RGB))
  width, height = image.size
  format = ultimateAlprSdk.ULTALPR_SDK_IMAGE_TYPE_RGB24

  result = ultimateAlprSdk.UltAlprSdkEngine_process(
      format,
      image.tobytes(), # type(x) == bytes
      width,
      height,
      0, # stride
      1
      )

  result = json.loads(result.json())
  output = []
  if "plates" in result.keys():
    for x in result["plates"]:
      output.append(x["text"])

  if "plates" in result.keys():
    print("plate:", result["plates"][0]["text"], flush=True)
    print("Recent visits:", recent_visits, flush=True)
    print("count:", last_plate_count, flush=True)
    if last_plate == result["plates"][0]["text"]:
      last_plate_count += 1
    else:
      last_plate_count = 0
    last_plate = result["plates"][0]["text"]

    if last_plate_count >= 30:
      if last_plate not in recent_visits.keys():
        add_plate(last_plate)
      elif (datetime.today() - recent_visits[last_plate]) > timedelta(minutes = 5):
        add_plate(last_plate)




  #print(output)
  #cv2.imshow("VIDEO", frame)
  #cv2.waitKey(1)

ultimateAlprSdk.UltAlprSdkEngine_deInit()


main.py
from datetime import datetime
import cv2
import numpy as np
import imutils
import easyocr

def picure(number):
    #define variable img for used image
    img = cv2.imread('/home/jouadamis/PycharmProjects/pythonProject/frame%d.jpg' % number)
    #img = cv2.imread('/home/jouadamis/PycharmProjects/pythonProject/image1.jpg')
    #converting picture to gray
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    #using bilateral Filter that reduces noise
    bfilter = cv2.bilateralFilter(gray, 11, 17, 17)
    #detecting edges
    edged = cv2.Canny(bfilter, 30, 200)
    #find edges
    keypoints = cv2.findContours(edged.copy(), cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
    #simplifying return of contours
    contours = imutils.grab_contours(keypoints)
    #return of top ten kontours
    contours = sorted(contours, key=cv2.contourArea, reverse=True)[:10]
    #filtring(looping) thrue contours to find number plate shape(rectangle)

    location = None
    for contour in contours:
        approx = cv2.approxPolyDP(contour, 12, True)
        if len(approx) == 4:
            location = approx
            break

    mask = np.zeros(gray.shape, np.uint8)
    #finding numberplate
    new_image = cv2.drawContours(mask, [location], 0, 255, -1)
    #overlaying everything except numberplate
    new_image = cv2.bitwise_and(img, img, mask=mask)
    #storing coordinates in variables x,y
    (x, y) = np.where(mask == 255)
    #storing top left point(min) and bottom right point(max)
    (x1, y1) = (np.min(x), np.min(y))
    (x2, y2) = (np.max(x), np.max(y))
    #getting variable croped_image
    cropped_image = gray[x1:x2+1, y1:y2+1]
    #define in what leanguage to read in
    reader = easyocr.Reader(['en'])
    #read text from variable cropped_image
    result = reader.readtext(cropped_image)
    #Render Result
    text1 = result[0][-2]
    text2 = result[1][-2]
    text = text1 + " " + text2
    #define font on picture
    font = cv2.FONT_HERSHEY_SIMPLEX
    #putting text in picture
    res = cv2.putText(img, text=text, org=(approx[0][0][0], approx[1][0][1]+80),
                      fontFace=font, fontScale=1, color=(0, 255, 0), thickness=2, lineType=cv2.LINE_AA)
    #open txt file
    f = open("demofile.txt", "a")
    #define date + time to variable time
    time = datetime.now()
    #write in file SPZ and date+time
    f.write(text + "    " + str(time) + " \n")
    #close file
    f.close()
    #drawing rectangle in picture
    res = cv2.rectangle(img, tuple(approx[0][0]), tuple(approx[2][0]), (0, 255, 0), 3)
    #save image

    cv2.imwrite('/home/jouadamis/PycharmProjects/pythonProject/image_new%d.jpg' % number,
               cv2.cvtColor(res, cv2.COLOR_BGR2RGB))
    #plt.imsave('image_new1.jpg', cv2.cvtColor(res, cv2.COLOR_BGR2RGB))


#define video
video = cv2.VideoCapture('/home/jouadamis/PycharmProjects/pythonProject/video1.mp4')
#read frame by frame
#define variables
success,image = video.read()
count = 0
frames = 1
skip_frames = 15
#loop function for using only frames % 31
while video.isOpened() and count < 3:
    ret, frame = video.read()
    if ret and frames % skip_frames == 0:
        # save frame as JPEG file
        cv2.imwrite("frame%d.jpg" % count, image)
        #add plus one to variables to use another frame and save next image with different name
        success, image = video.read()
        print(count)
        picure(count)
        count += 1
        frames += 1

    else:
        frames += 1

video.release()

print("done :)")
